File: droid/app/src/main/java/com/storyteller/droid/model/SaveState.kt
package com.storyteller.droid.model

import com.google.gson.Gson
import java.io.File
import java.nio.file.Files
import java.nio.file.StandardCopyOption
import java.util.UUID

data class ChatTurn(val role: String, val text: String)

/** App-private state. Procedural/package facts are referenced, never copied. */
data class SaveState(
    val saveVersion: Int = 1,
    val storyId: String = "",
    val packageContentHash: String = "",
    val playthroughId: String = UUID.randomUUID().toString(),
    var currentNode: String = "",
    val visitedNodes: MutableList<String> = mutableListOf(),
    val flags: MutableMap<String, Boolean> = mutableMapOf(),
    val bookmarks: MutableList<String> = mutableListOf(),
    val gmHistory: MutableList<ChatTurn> = mutableListOf(),
    val choiceHistory: MutableList<String> = mutableListOf(),
) {
    var currentNodeId: String
        get() = currentNode
        set(value) { currentNode = value }

    fun visitNode(nodeId: String) { currentNode = nodeId; if (nodeId !in visitedNodes) visitedNodes += nodeId }
    fun makeChoice(choice: Choice) { choiceHistory += choice.choiceId; choice.setsFlags.forEach { flags[it] = true } }
    fun addGmExchange(question: String, answer: String) {
        gmHistory += ChatTurn("user", question); gmHistory += ChatTurn("assistant", answer)
    }
    fun toggleBookmark(): Boolean = if (bookmarks.remove(currentNode)) false else { bookmarks += currentNode; true }
    fun reset(entryNode: String = visitedNodes.firstOrNull() ?: currentNode) {
        currentNode = entryNode; visitedNodes.clear(); visitedNodes += entryNode
        flags.clear(); choiceHistory.clear(); gmHistory.clear(); bookmarks.clear()
    }
    fun save(saveDir: File) = SaveRepository(saveDir.parentFile ?: saveDir)
        .saveAt(this, File(saveDir, "save_state.json"))

    companion object {
        /** Compatibility adapter; new code uses [SaveRepository]. */
        fun load(saveDir: File): SaveState = SaveRepository(saveDir.parentFile ?: saveDir)
            .loadAny(saveDir) ?: SaveState()
    }
}

class SaveHashMismatch : IllegalStateException("SAVE_PACKAGE_HASH_MISMATCH")

class SaveRepository(private val root: File) {
    private val gson = Gson()
    private fun file(storyId: String, playthroughId: String) =
        File(root, "saves/$storyId/$playthroughId.json")

    fun load(story: StoryPackage, playthroughId: String): SaveState? {
        val path = file(story.storyId, playthroughId)
        if (!path.isFile) return null
        val state = gson.fromJson(path.readText(), SaveState::class.java)
        if (state.storyId != story.storyId || state.packageContentHash != story.contentHash) throw SaveHashMismatch()
        return state
    }

    fun save(state: SaveState) = saveAt(state, file(state.storyId, state.playthroughId))

    internal fun saveAt(state: SaveState, destination: File) {
        destination.parentFile?.mkdirs()
        val temp = File(destination.parentFile, ".${destination.name}.tmp")
        temp.outputStream().use { stream ->
            stream.write(gson.toJson(state).toByteArray(Charsets.UTF_8)); stream.fd.sync()
        }
        Files.move(temp.toPath(), destination.toPath(), StandardCopyOption.REPLACE_EXISTING,
                   StandardCopyOption.ATOMIC_MOVE)
    }

    internal fun loadAny(directory: File): SaveState? {
        val path = File(directory, "save_state.json")
        return if (path.isFile) runCatching { gson.fromJson(path.readText(), SaveState::class.java) }.getOrNull() else null
    }

    fun deleteStoryData(storyId: String) = File(root, "saves/$storyId").deleteRecursively()
}

File: ios/StoryTeller/Data/SaveState.swift
import Foundation

struct ChatTurn: Codable, Equatable { let role: String; let text: String }

struct SaveState: Codable, Equatable {
    let saveVersion: Int
    let storyId: String
    let packageContentHash: String
    let playthroughId: String
    var currentNode: String
    var visitedNodes: [String]
    var flags: [String: Bool]
    var bookmarks: [String]
    var gmHistory: [ChatTurn]
    var choiceHistory: [String]
    var currentNodeId: String { get { currentNode } set { currentNode = newValue } }

    init(storyId: String = "", packageContentHash: String = "", playthroughId: String = UUID().uuidString,
         currentNode: String = "", visitedNodes: [String] = [], flags: [String: Bool] = [:],
         bookmarks: [String] = [], gmHistory: [ChatTurn] = [], choiceHistory: [String] = []) {
        saveVersion=1; self.storyId=storyId; self.packageContentHash=packageContentHash
        self.playthroughId=playthroughId; self.currentNode=currentNode; self.visitedNodes=visitedNodes
        self.flags=flags; self.bookmarks=bookmarks; self.gmHistory=gmHistory; self.choiceHistory=choiceHistory
    }
    mutating func visitNode(_ id: String) { currentNode=id; if !visitedNodes.contains(id) { visitedNodes.append(id) } }
    mutating func makeChoice(_ choice: Choice) { choiceHistory.append(choice.choiceId); choice.setsFlags.forEach { flags[$0]=true } }
    mutating func addGMExchange(question: String, answer: String) {
        gmHistory += [ChatTurn(role:"user", text:question), ChatTurn(role:"assistant", text:answer)]
    }
    mutating func toggleBookmark() -> Bool {
        if let index=bookmarks.firstIndex(of:currentNode) { bookmarks.remove(at:index); return false }
        bookmarks.append(currentNode); return true
    }
    mutating func reset(entryNode: String? = nil) {
        let entry=entryNode ?? visitedNodes.first ?? currentNode
        currentNode=entry; visitedNodes=[entry]; flags=[:]; bookmarks=[]; gmHistory=[]; choiceHistory=[]
    }
    static func load(from directory: URL) -> SaveState {
        (try? SaveRepository(root: directory.deletingLastPathComponent()).loadAny(directory)) ?? SaveState()
    }
    func save(to directory: URL) { try? SaveRepository(root: directory.deletingLastPathComponent()).saveAt(self, directory.appendingPathComponent("save_state.json")) }
}

enum SaveRepositoryError: Error, Equatable { case packageHashMismatch }

struct SaveRepository {
    let root: URL
    private let fm = FileManager.default
    private func url(_ story: String, _ playthrough: String) -> URL {
        root.appendingPathComponent("saves/\(story)/\(playthrough).json")
    }
    func load(story: StoryPackage, playthroughId: String) throws -> SaveState? {
        let path=url(story.storyId, playthroughId); guard fm.fileExists(atPath:path.path) else{return nil}
        let state=try JSONDecoder().decode(SaveState.self, from:Data(contentsOf:path))
        guard state.storyId==story.storyId && state.packageContentHash==story.contentHash else{throw SaveRepositoryError.packageHashMismatch}
        return state
    }
    func save(_ state: SaveState) throws { try saveAt(state, url(state.storyId,state.playthroughId)) }
    func saveAt(_ state: SaveState, _ destination: URL) throws {
        try fm.createDirectory(at:destination.deletingLastPathComponent(),withIntermediateDirectories:true)
        let data=try JSONEncoder().encode(state); let temp=destination.deletingLastPathComponent().appendingPathComponent(".\(destination.lastPathComponent).tmp")
        try data.write(to:temp,options:.atomic); if fm.fileExists(atPath:destination.path){_ = try fm.replaceItemAt(destination,withItemAt:temp)}else{try fm.moveItem(at:temp,to:destination)}
    }
    func loadAny(_ directory: URL) throws -> SaveState? {
        let path=directory.appendingPathComponent("save_state.json"); guard fm.fileExists(atPath:path.path) else{return nil}
        return try JSONDecoder().decode(SaveState.self,from:Data(contentsOf:path))
    }
    func deleteStoryData(_ storyId:String)throws{let p=root.appendingPathComponent("saves/\(storyId)");if fm.fileExists(atPath:p.path){try fm.removeItem(at:p)}}
}
